x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/optimizer/Optimizer.java

```diff
@@ -122,4 +126,31 @@ protected LogicalPlan rule(Filter filter) {
             });
         }
     }
+
+    private static class MatchLiteralsOnTheRight extends OptimizerRule<Filter> {
+
+        @Override
+        protected LogicalPlan rule(Filter filter) {
+            return filter.transformExpressionsUp(e -> {
+                if (e instanceof Match) {
+                    Match m = (Match) e;
+                    if (m.children().get(0).foldable()) {
+                        int size = m.children().size();
+                        List<Expression> newChildren = new ArrayList<>(size);
+                        Expression field = null;
+                        for (Expression c : m.children()) {
+                            if (c.foldable() == false) {
+                                field = c;
+                            } else {
+                                newChildren.add(c);
+                            }
+                        }
+                        newChildren.add(0, field);
+                        return m.replaceChildren(newChildren);
```

Comment:
Why isn't this done inside the main `Match` constructor or even better inside `makeSubstitute`?